go/pools/fast_pool.go

```diff
@@ -0,0 +1,659 @@
+func (p *FastPool) recordWait(startTime time.Time) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.state.WaitCount++
+	p.state.WaitTime += time.Now().Sub(startTime)
+}
+
+// usedFromPool moves a resource from `InPool` to `InUse`.
+func (p *FastPool) usedFromPool() {
+	defer p.Unlock()
+	p.Lock()
+	p.state.InPool--
+	p.state.InUse++
+}
+
+// Put back a resource into the pool. For every successful Get,
+// a corresponding Put is required. If you no longer need a resource,
+// you will need to call Put(nil) instead of returning the closed resource.
+// The will eventually cause a new resource to be created in its place.
+func (p *FastPool) Put(resource Resource) {
+	p.Lock()
+
+	p.state.InUse--
+
+	if p.state.Closed || p.active() > p.state.Capacity {
+		// We're ether closed or shrinking.
+		p.Unlock()
+		if resource != nil {
+			resource.Close()
+		}
+		return
+	}
+
+	defer p.Unlock()
+
+	if resource == nil {
+		return
+	}
+
+	if p.state.InUse < 0 {
+		p.state.InUse++
+		panic(ErrPutBeforeGet)
+	}
+
+	w := resourceWrapper{resource: resource, timeUsed: time.Now()}
+	select {
+	case p.pool <- w:
+		p.state.InPool++
+	default:
+		// We don't have room in the pool.
+		p.state.InUse++
+		panic(ErrFull)
+	}
+}
+
+// SetCapacity changes the capacity of the pool.
+// You can use it to shrink or expand, but not beyond
+// the max capacity. If the change requires the pool
+// to be shrunk and `block` is true, SetCapacity waits
+// till the necessary number of resources are returned
+// to the pool.
+//
+// A SetCapacity of 0 is equivalent to closing the pool.
+func (p *FastPool) SetCapacity(capacity int, block bool) error {
+	err := p.setCapacityHandle(capacity)
+	if err != nil {
+		return err
+	}
+	if !p.State().Draining {
+		return nil
+	}
+
+	if block {
+		p.shrink()
+	} else {
+		go p.shrink()
+	}
+
+	return nil
+}
+
+// setCapacityHandle is used by SetCapacity to do sanity checks and change
+// state appropriately, while locking the state. It will also let the caller know
+// if the pool needs to shrink.
+func (p *FastPool) setCapacityHandle(capacity int) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.state.Closed {
+		return ErrClosed
+	}
+	if capacity < 0 || capacity > cap(p.pool) {
+		return fmt.Errorf("capacity %d is out of range", capacity)
+	}
+	if capacity > 0 {
+		minActive := p.state.MinActive
+		if capacity < minActive {
+			return fmt.Errorf("minActive %v would now be higher than capacity %v", minActive, capacity)
+		}
+	}
+	if capacity == 0 {
+		p.state.Closed = true
+	}
+
+	p.state.Draining = capacity < p.state.Capacity
+	p.state.Capacity = capacity
+	return nil
+}
+
+// shrink active resources until the active count is not above capacity.
+func (p *FastPool) shrink() {
+	for p.continueShrink() {
+		// We can't remove InUse resources, so only target the pool.
+		// Collect the InUse resources lazily when they're returned.
+		select {
+		case wrapper := <-p.pool:
+			wrapper.resource.Close()
+			wrapper.resource = nil
+
+			p.withLock(func() {
```

Comment:
Getting a resource from the pool and later decrementing `InPool` means that you cannot rely on `InPool` to accurately represent what's in the pool. This means that any logic that checks `InPool` and then performs an action is a potential bug. It looks like `active` depends on it, and is used in many places.